[edures888] iP

edures888 opened the pull request:
## **[ModifiedDuke](https://www.oracle.com/java/duke/)**
> “Any program is only as good as it is useful.” by Linus Torvalds.

- [x] Minimal Goals
- [ ] Adding JavaFX GUI
- [ ] Stretch Goals and more Java features
- [ ] Testing and Documentation



* Simple to use 👍 
* Tasks with additional info
   1. Deadlines
   2. Beginning and end of events
 
It all started from `Duke.java` here:
```java
public class Duke {
    public static void main(String[] args) {
        String logo = " ____        _        \n"
                + "|  _ \\ _   _| | _____ \n"
                + "| | | | | | | |/ / _ \\\n"
                + "| |_| | |_| |   <  __/\n"
                + "|____/ \\__,_|_|\\_\\___|\n";
        System.out.println("Hello from\n" + logo);
    }
}
```

Commit:
Add Gradle support

edures888 committed:
Implement Level 1 functionality

edures888 committed:
Level-2 implementation

edures888 committed:
Level-3 implementation

edures888 committed:
Level-4 Implementation

edures888 committed:
A-TextUiTesting Implementation

edures888 committed:
Level-5

edures888 committed:
Level-6

edures888 committed:
Level 7. Save

Currently, tasks are not persistent upon exiting Duke.

Save the tasks in the hard disk automatically whenever the task list changes.
Load the data from the hard disk when Duke starts up.

edures888 committed:
Level 8. Dates and Times

Dates in Deadline tasks and Event tasks were stored as Strings and
displayed in the same format

Changing to Date objects could allow for customized display of the dates
and even easier comparison between them

edures888 committed:
Level 8. Dates and Times

Dates in Deadline tasks and Event tasks were stored as Strings and
displayed in the same format

Changing to Date objects could allow for customized display of the dates
and even easier comparison between them

Change type stored for dates to be Date object and also update the
string representations accordingly

edures888 committed:
Merge branch 'branch-Level-7'
Extend to level 7 functionality

edures888 committed:
Merge branch 'branch-Level-8'
Extend functionality to Level 8

edures888 committed:
Changed gitignore to exclude saved tasks

edures888 committed:
WIP A-OOP

edures888 committed:
Merge remote-tracking branch 'origin/add-gradle-support'
A-Gradle

edures888 committed:
A-Gradle

edures888 committed:
WIP

edures888 committed:
WIP A-OOP

edures888 committed:
Fix gradle build file

edures888 committed:
Add JUnit tests to test the behavior of ListCommand and DeadlineTask

edures888 committed:
Improve code quality by following coding standards

Add checkstyle to gradle build and made the changes accordingly.
Improve code readability by adding some comments and extra line breaks between
units of logic.

edures888 committed:
Add JavaDocs in progress

edures888 committed:
Add JavaDoc header comment for all classes

edures888 committed:
Implement task search functionality

Extend the Task class and also make a new command class for the search
command.

edures888 committed:
Fix merge conflict between branch-A-CodingStandard and branch-A-JavaDoc

edures888 committed:
Fix small merge conflict when merging branch-Level-9 into master

edures888 committed:
Add new search command result

Initially, list command result was also used for search command

Add new search command in place of that

edures888 committed:
Add more JavaDoc header comments for classes and methods

edures888 committed:
Add JavaFX implementation for GUI in progress

Wong-Jia-Jun commented on src/test/java/duke/commmands/ListCommandTest.java:
```diff
@@ -0,0 +1,24 @@
+public class ListCommandTest {
+    static final String NORMAL_INPUT = "list";
+    static final String SIMPLE_TASK_NAME = "simpleTask";
+    static final String CORRECT_TASK_OUTPUT = "Current tasks in list:\n1. [Todo] [ ] simpleTask";
```
I think using Java enums will be good  to abstract the concept of commands. e.g list,delete etc

Wong-Jia-Jun commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+
+    private ArrayList<Task> list;
+    public TaskList() {
+        list = new ArrayList<>();
```
Just nitpicking a little but could name this ArrayList variable more comprehensively e.g. listOfTasks

edures888 replied:
Right, that's a good catch. Thanks!

Wong-Jia-Jun reviewed (commented):
Overall i think it looks good, works and follows coding standards.

edures888 committed:
Add GUI to Duke using JavaFX

Was using CLI for user interface

Add JavaFX functionality by adding launcher class, GUI component
controllers and respective XML files, and then refactor code to remove
UserInterface class. Transfer Controller class functionality to Duke
class.

daytona65 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,32 @@
+        parser = new Parser();
+    }
+
+    public String executeCommand(String input) throws DukeException {
```
Missing Javadocs

daytona65 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,32 @@
+ */
+public class Duke {
+    public static final String TERMINATION_INDICATION = "TERMINATE";
+    private TaskList tasks;
```
I like how you differentiated singular vs plural variables in your variable naming!

daytona65 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,16 @@
+ * General exception for the Duke command for displaying messages to user
+ */
+public class DukeException extends Exception{
+    private static final String CUSTOM_PREFIX = "Something's not right.\n";
```
Follows naming standards, good work

daytona65 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,72 @@
+     * @return List of tasks
+     * @throws DukeException If save file path or save file itself is invalid
+     */
+    public TaskList recoverList() throws DukeException {
```
Method names are good and concise, letting me know what it does!

daytona65 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+ */
+public class TaskList implements Serializable {
+
+    private static final String INVALID_INDEX_EXCEPTION = "Invalid task index given for Mark/Unmark/Delete command.";
+    private static final String ADDED_TASK_RESPONSE = "Task added:\n";
+    private static final String REMAINING_TASK_RESPONSE = "\nRemaining task count: ";
+    private static final String REMOVE_TASK_RESPONSE = "Task removed:\n";
+    private static final String LIST_INDEX_SEPARATOR = ". ";
+
```
You can consider using enums or a string array to store these final values

daytona65 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+
+    public String unmarkTask(int taskIndex) throws DukeException {
+        try {
+            return list.get(taskIndex - 1).unmark();
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException(INVALID_INDEX_EXCEPTION);
+        }
+    }
+
+    public String deleteTask(int taskIndex) throws DukeException {
+        try {
+            Task toRemove = list.get(taskIndex - 1);
+            list.remove(taskIndex - 1);
+            return REMOVE_TASK_RESPONSE + toRemove.toString() + REMAINING_TASK_RESPONSE + list.size();
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException(INVALID_INDEX_EXCEPTION);
+        }
+    }
+
+    public String search(String keyword) {
+        TaskList filteredTasks = new TaskList();
+        for (Task task: list) {
+            if (task.hasKeyword(keyword)) {
+                filteredTasks.addTask(task);
+            }
+        }
+        return filteredTasks.toString();
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder result = new StringBuilder();
+
+        for (int i = 0; i < list.size(); i++) {
+            result.append("\n").append(i + 1).append(LIST_INDEX_SEPARATOR).append(list.get(i));
+        }
+        return result.toString();
+    }
+
+}
```
Missing Javadocs

daytona65 reviewed (commented):
Overall good work, code is clean and largely follows coding standards. Keep it up!

edures888 committed:
Refactor small detail

Variable name in TaskList may be ambiguous

Rename variable to be more descriptive

Readers may better understand the purpose and context of the variable

edures888 committed:
Refactor to improve quality

Some variables have short abbreviated names without much information.
Some magic strings are present as well.

Let's change the variable names to be fully spelt out and be more
descriptive. Assign magic strings to constants represented by
class-level members.

Readers can get a clearer idea of what each of the variables mean.
Readers can further understand the meaning behind the string constants
used. Future changes may be more convenient as one would only have to
alter the constants, instead of having to change them at all of the
locations they are used at.

edures888 committed:
Add assert statements

Conditionals and try-catch blocks are being used without any assert
statements.

Readers may overlook some necessary assumptions or even question the
absence of handlers for the respective edge cases violating those
assumptions.

Let's add assert statements in relevant parts of the code.

Using assert statements can signal additional information to reader on
the numerous assumptions at the respective parts of the code.

edures888 committed:
Add usage of Java streams

edures888 committed:
Merge pull request #2 from edures888/branch-A-Streams

Add usage of Java streams

edures888 committed:
Merge pull request #3 from edures888/branch-A-Assertions

Add assert statements

edures888 committed:
Merge pull request #4 from edures888/branch-A-CodeQuality

Refactor to improve quality

edures888 committed:
Setup CI function with Github Actions

edures888 committed:
Add increment for more flexibility in searching for items

Items can be found even if the keyword matches the item only partially

edures888 committed:
Refactor to pass checkystyle tests

edures888 committed:
Add GUI termination functionality

edures888 committed:
Fix fxml variable id bug

edures888 committed:
Fix checkstyle errors

edures888 committed:
Fix javadocs issues

edures888 committed:
Fix javafx dependency issue in gradle file

edures888 committed:
Add User Guide

edures888 committed:
Fix text overrun issues in FXML

edures888 committed:
Fix line separator bug for test

edures888 committed:
Fix checkstyle error
